src/controllers/orderController.js

```diff
@@ -5,58 +5,103 @@ const { BadRequestError } = require('../errors');
-  }
 
-  // Create an Order document in the database
-  const order = await Order.create({
-    user: req.user.userId,
-    orderItems,
-    subtotal,
-    tax,
-    total,
-  });
 
-  // Create a payment intent with Stripe
-  const paymentIntent = await stripe.paymentIntents.create({
-    amount: total * 100, // Stripe requires amount in cents
-    currency: 'usd',
-    description: `Order #${order._id}`,
-    metadata: {
-      userId: req.user.userId,
-      orderId: order._id, // key-values
-      totalQuantity: orderItems.reduce(
-        (total, item) => total + item.quantity,
-        0
-      ), // the total number of items in the order
-      totalAlbums: orderItems.length, // the total number of albums purchased in order
-    },
-  });
 
-  // Save the paymentIntent to the Order model
-  order.paymentIntentId = paymentIntent.id;
-  await order.save();
 
-  // Create PurchasedAlbum entries for each album in orderItems
-  for (const orderItem of orderItems) {
-    await PurchasedAlbum.create({
-      album: orderItem.album, // album ID from orderItem
-      user: req.user.userId, // user ID from the request
+// Populate the orderItems with full album data
+const populatedOrderItems = await Promise.all(
+  orderItems.map(async (orderItem) => {
+    return await orderItem.populate('album').execPopulate();
+  })
+);
+  // Create an array to store the order items with full album data
+  const orderItemsWithFullAlbum = populatedOrderItems.map((orderItem) => ({
+    album: orderItem.album.toObject(), // Use toObject() to convert the full album to a plain JavaScript object
+    quantity: orderItem.quantity,
+  }));
+
+
+
+
+    const order = new Order({
+      user: req.user.userId,
+      orderItems: orderItemsWithFullAlbum,// Use the array with full album data
+      subtotal,
+      tax,
+      total,
     });
-  }
 
-  // Send the payment intent client secret and order information to the client
-  res
-    .status(StatusCodes.CREATED)
-    .json({ clientSecret: paymentIntent.client_secret, order });
+     await order.save({ session });
+
+   
+    const paymentIntent = await stripe.paymentIntents.create({
+      amount: total * 100,
+      currency: 'usd',
+      description: `Order #${order._id}`,
+      metadata: {
+        userId: req.user.userId,
+        orderId: order._id,
+        totalQuantity: orderItems.reduce(
+          (total, item) => total + item.quantity,
+          0
+        ),
+        totalAlbums: orderItems.length,
+      },
+    });
+
+    order.paymentIntentId = paymentIntent.id;
+
+    for (const orderItem of orderItems) {
+      const purchasedAlbum = new PurchasedAlbum({
+        album: orderItem.album,
+        user: req.user.userId,
+      });
+
+      await purchasedAlbum.save({ session });
+    }
+
+
+
+
+    await session.commitTransaction();
+
+    
+    // Send the order completion email after THE ORDER HAS STATUS COMPLETE
+    await sendOrderCompletedEmail(user.email, user.name, orderItemsWithFullAlbum);
```

Comment:
Here we could send an email that says "your order is being processed"

but if we only want to send an email when the order status is complete, then instead of putting this `sendOrderCompletedEmail` here then a better place for now would be to put it into a `.post('save',...)` middleware in the Order model file. Or in an `updateOrder` endpoint that the frontend can call when the checkout process has been completed. I will also include a ticket for listening to a webhook from stripe directly for this usecase.

Reply:
@akosasante created a `.post('save',...) middleware ` but I've been testing and it doesn't send email when the orderStatus is changed to 'complete' manually (directly in mongo db)- so i added he call of this save() middleware on line 62 in orderController.js -but it also doesn't help to send an email...

Reply:
Changing the status directly in MongoDB will _**not**_ trigger any mongoose middleware, because mongoose is a separate library from the MongoDB itself.

Reply:
We can either (1) use a node script to update the `orderStatus` of an order to `complete` using mongoose (Order.update) (just for dev testing purposes), or (2) add an `updateOrder` endpoint/controller function that we can use to send an update request via Postman, or (3) update to `complete` based on Stripe webhook events

Reply:
@akosasante I took https://codethedreamteam4-2023.atlassian.net/browse/PS-202 , a 3d option